modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp

```diff
@@ -13,25 +13,55 @@
 // limitations under the License.
 
 #include "contextwrapper.hpp"
+#include <execution_graph/Context.h>
+#include <node_cudf/table.hpp>
+#include "cache.hpp"
 
 namespace nv {
 
-Napi::Function ContextWrapper::Init(Napi::Env env, Napi::Object exports) {
+Napi::Function ContextWrapper::Init(Napi::Env const& env, Napi::Object exports) {
   return DefineClass(env, "ContextWrapper", {});
 }
 
 ContextWrapper::wrapper_t ContextWrapper::New(
   Napi::Env const& env,
+  int32_t const& ral_id,
   std::pair<
     std::pair<std::shared_ptr<ral::cache::CacheMachine>, std::shared_ptr<ral::cache::CacheMachine>>,
-    int> args) {
-  auto inst    = EnvLocalObjectWrap<ContextWrapper>::New(env, {});
-  auto& caches = args.first;
-  inst->_port  = args.second;
+    int> const& pair,
+  Napi::Object const& ucp_context) {
+  auto inst            = EnvLocalObjectWrap<ContextWrapper>::New(env, {});
+  auto& caches         = pair.first;
+  inst->_port          = pair.second;
+  inst->_ral_id        = ral_id;
+  inst->_transport_in  = Napi::Persistent(CacheMachine::New(env, caches.second));
+  inst->_transport_out = Napi::Persistent(CacheMachine::New(env, caches.first));
+  inst->_ucp_context   = Napi::Persistent(ucp_context);
   return inst;
 }
 
 ContextWrapper::ContextWrapper(Napi::CallbackInfo const& info)
   : EnvLocalObjectWrap<ContextWrapper>(info) {}
 
+void ContextWrapper::add_to_cache(int32_t const& node_id,
+                                  int32_t const& src_ral_id,
+                                  int32_t const& dst_ral_id,
+                                  std::string const& ctx_token,
+                                  std::string const& message_id,
+                                  std::vector<std::string> const& column_names,
+                                  cudf::table_view const& table_view) {
+  if (src_ral_id == dst_ral_id) {
+    this->_transport_in.Value()->add_to_cache(
+      node_id, src_ral_id, dst_ral_id, ctx_token, message_id, column_names, table_view);
+  } else {
+    this->_transport_out.Value()->add_to_cache(
+      node_id, src_ral_id, dst_ral_id, ctx_token, message_id, column_names, table_view);
+  }
```

Comment:
I forgot to delete this after testing -- a BlazingContext can't send messages to itself.

```suggestion
  this->_transport_out.Value()->add_to_cache(
    node_id, src_ral_id, dst_ral_id, ctx_token, message_id, column_names, table_view);
```